### PR TITLE
Implement S.N.NameResolution for Linux and OS X.

### PR DESCRIPTION
**Pull request by pgavlin**

### CODE REVIEW

**Comment by pgavlin** [comment]
All tests pass on Windows; I verified the *nix functionality manually.

File: src/System.Net.NameResolution/src/System.Net.NameResolution.csproj [hunk 1]
```diff
@@ -5,26 +5,22 @@
     <AssemblyName>System.Net.NameResolution</AssemblyName>
     <AssemblyVersion>4.0.0.0</AssemblyVersion>
     <OutputType>Library</OutputType>
-    <ProjectGuid>{1714448C-211E-48C1-8B7E-4EE667D336A1}</ProjectGuid>
+    <ProjECtGuiD>{F6D1C093-081D-46DE-B5A8-516533375FDD}</ProjECtGuiD>
```
Comment (CIPop) [inline src/System.Net.NameResolution/src/System.Net.NameResolution.csproj]: @pgavlin Any reason to change the Guid? Also, note the capitalization issue.

Comment (pgavlin) [inline src/System.Net.NameResolution/src/System.Net.NameResolution.csproj]: Ack, I meant to change the GUID in the test project. Will fix.

File: src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs [hunk 2]
```diff
@@ -14,8 +15,8 @@ internal static class NameResolutionPal
         // used by GetHostName() to preallocate a buffer for the call to gethostname.
         //
         private const int HostNameBufferLength = 256;
-        private static volatile bool s_Initialized;
-        private static object s_InitializedLock = new object();
+        private static bool s_initialized;
```
Comment (CIPop) [inline src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs:18]: @pgavlin The https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md isn't clear on this: should we not capitalize after s_?

File: src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs [hunk 3]
```diff
@@ -362,11 +367,11 @@ public static string GetHostName()
 
         public static void EnsureSocketsAreInitialized()
         {
-            if (!s_Initialized)
+            if (!Volatile.Read(ref s_initialized))
```
Comment (CIPop) [inline src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs:370]: @pgavlin Thanks! Took a note to update other places where we have this pattern.

File: src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs (hunk 2 above)
Comment (pgavlin) [inline src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs:18]: It seems to be an ad-hoc convention that the `s_` is not treated as part of the field name for the purposes of camel-casing.

Comment (pgavlin) [inline src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs:18]: (also, I'm reasonably sure that CodeFormatter will rename a private static field called e.g. `SomePrivateStatic` to `s_somePrivateStatic`)

Comment (CIPop) [inline src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs:18]: @pgavlin If the tool already lower-cases the first letter after `s_`, then we should update the docs to add an example. I don't have a preference for casing but I like consistency. Took a note to track this.

**Comment by CIPop** [comment]
:+1:

**Pull request merged**